### PR TITLE
Discover tests via LSP if available

### PR DESCRIPTION
**Pull request by plemarquand**
If the LSP is available and supports the two requests for listing tests then use those to discover tests in the workspace/document. If these requests are unsupported or if they fail, fall back to the existing methods.

This patch uses the existing code to leverage the LSP's workspace/tests method and adds the textDocument/tests for listing tests within documents as they change.

These requests produce a collection of TestClass structures, which are hierarchical. These are passed to one of the methods in TestDiscovery.ts which diffs against the existing tree and adds/removes tests. This is similar to how it already worked, but now supports arbitrary levels of nesting in the TestClass structure.

This patch also adds preliminary support for listing swift-testing tests through the LSP methods, as well as through `swift test --list-tests`. It does not support finding swift-testing tests through the document symbols.

Another small improvement is if listing tests fails via `swift test --list-tests` we'll attempt a `swift build --build-tests` before trying again. Today, if the user was on the 5.10 toolchain and no build had yet been performed they would get a somewhat cryptic error in the test explorer.

Swift-testing tests are filtered out of the list of available tests until code is added to run them. This will be added with #757.

- Adds tests for parsing the results of `swift test --list-tests`
- Adds tests for parsing tests via document symbols

This patch addresses both #761 and #762.

### CODE REVIEW

**Comment by plemarquand** [comment]
Note: This will only work with the latest sourcekit-lsp. To test you can clone the sourcekit-lsp repo, `swift build` and then set the `swift.sourcekit-lsp.serverPath` VSCode setting to point to your built sourcekit-lsp.

File: src/TestExplorer/LSPTestDiscovery.ts
Comment (award999) [inline src/TestExplorer/LSPTestDiscovery.ts:79]: todo: "workspace/tests"

suggestion: Since same logic for textDocument, we could extract this into it's own function so we can re-use if more experimental capabilities are added

File: src/TestExplorer/LSPTestDiscovery.ts
Comment (award999) [inline src/TestExplorer/LSPTestDiscovery.ts:139]: suggestion: maybe we fetch this list in "transform" so we're not filtering each time the map callback is called

Comment (plemarquand) [reply src/TestExplorer/LSPTestDiscovery.ts]: Any given test might be in a different test target, based on its location. This means we need to check each test's target. A cached lookup of locations to test targets may improve performance a bit but there wont likely be a lot of test targets to look through in a given project.

File: src/TestExplorer/LSPTestDiscovery.ts
Comment (plemarquand) [reply src/TestExplorer/LSPTestDiscovery.ts]: Good call, I've refactored out the common capability check into a method.

**Review by award999 — COMMENTED** [review]
looks good to me, but probably should get another set of eyes

File: src/TestExplorer/LSPTestDiscovery.ts
Comment (adam-fowler) [inline src/TestExplorer/LSPTestDiscovery.ts:99]: I did not know about this. It would be good if Microsoft actually documented the language client APIs. Is the availability of the workspace/tests and textDocument/tests something we can cache?

File: src/TestExplorer/LSPTestDiscovery.ts
Comment (adam-fowler) [inline src/TestExplorer/LSPTestDiscovery.ts:110]: This seems like a very generic method name. `transformToTestClass` maybe?

File: src/TestExplorer/TestExplorer.ts
Comment (adam-fowler) [inline src/TestExplorer/TestExplorer.ts:149]: Do we want to run this on every document symbols response, or when the active file is saved?
EDIT: although I see what you are doing, you are falling back to document symbols if LSP fails

File: src/TestExplorer/TestDiscovery.ts
Comment (adam-fowler) [inline src/TestExplorer/TestDiscovery.ts:88]:
```suggestion
    // Add/update the top level test items. upsertTestItem will descend the tree of children adding them as well.
```

File: src/TestExplorer/LSPTestDiscovery.ts
Comment (adam-fowler) [inline src/TestExplorer/LSPTestDiscovery.ts:75]: Have you found a situation where the tests returned weren't in the workspace? Also `updateTestsFromClasses` does this for individual workspace folders later on

File: src/TestExplorer/TestExplorer.ts
Comment (adam-fowler) [inline src/TestExplorer/TestExplorer.ts:151]: I guess for swift-testing we will need to remove this condition given `swift-testing` tests can be found in any target.

**Review by adam-fowler — COMMENTED** [review]
This looks great. Left a few comments about minor issues

File: src/TestExplorer/LSPTestDiscovery.ts
Comment (plemarquand) [reply src/TestExplorer/LSPTestDiscovery.ts:99]: The capabilities are actually cached on the `client` right after the connection is established to the server.

Comment (adam-fowler) [reply src/TestExplorer/LSPTestDiscovery.ts:99]: I actually meant caching in a boolean, instead of running this function. It's a small amount of code, but it gets run with every call

File: src/TestExplorer/TestExplorer.ts
Comment (plemarquand) [reply src/TestExplorer/TestExplorer.ts:149]: I think we want to run this on every response, as I'll often be typing out a new test declaration and it will show up immediately instead of waiting until save.

And yep, if `getDocumentTests` fails because the LSP isn't supported we fall back to the old way.

Comment (adam-fowler) [reply src/TestExplorer/TestExplorer.ts:149]: I guess `textDocument/tests` shouldn't be too heavy a request for the LSP server

File: src/TestExplorer/LSPTestDiscovery.ts
Comment (plemarquand) [reply src/TestExplorer/LSPTestDiscovery.ts:75]: I think you're right. I kept this filter from the old code but I dont think the LSP can return tests outside the workspace it was initialized with. I'll remove it.

File: src/TestExplorer/LSPTestDiscovery.ts
Comment (plemarquand) [reply src/TestExplorer/LSPTestDiscovery.ts:96]: Ah ok, I've cached the result in a Map keyed off the capability name + version.

File: src/TestExplorer/TestExplorer.ts
Comment (grynspan) [reply src/TestExplorer/TestExplorer.ts:151]: At this time, we would only want to support running tests in test targets. We document that requirement [here](https://swiftpackageindex.com/apple/swift-testing/0.7.0/documentation/testing/definingtests#app-top) (see the first **Note** callout.)

**Review by adam-fowler — APPROVED** [review]

**Comment by adam-fowler** [comment]
Before merging this I thought I'd test swift 5.10 and it looks like the document symbols loading isn't working

**Comment by adam-fowler** [comment]
Ah I see what you have done. By converting to the `swift test list` output, you lose the location information that comes with the document symbols. This will need to be fixed before this PR is merged

**Pull request merged**
